### PR TITLE
Fix #454: Port FxCop rule CA1721

### LINKED ISSUE #454: Port FxCop rule CA1721: PropertyNamesShouldNotMatchGetMethods
**Title:** Property names should not match get methods

**Description:**

The name of a public or protected member starts with ""Get"" and otherwise matches the name of a public or protected property. ""Get"" methods and properties should have names that clearly distinguish their function.

**Dependency:** None

**Notes:**

See if StyleCop has implemented it

### PR DESCRIPTION
**Pull request by morrisjoe**
@srivatsn @mavasani

### CODE REVIEW

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PropertyNamesShouldNotMatchGetMethods.cs [hunk 1]
```diff
@@ -15,35 +15,69 @@ public abstract class PropertyNamesShouldNotMatchGetMethodsAnalyzer : Diagnostic
         internal const string RuleId = "CA1721";
```
Comment (srivatsn) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PropertyNamesShouldNotMatchGetMethods.cs]: Can you make this classes sealed and add the `[DiagnsoticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]` attribute to it and then delete the stubs for the c#\vb specific types and the fixers.

Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PropertyNamesShouldNotMatchGetMethods.cs]: yes. re-submitted with changes to address this

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PropertyNamesShouldNotMatchGetMethods.cs [hunk 2]
```diff
@@ -6,44 +6,112 @@
...
+        private static readonly LocalizableString s_localizableBaseTypeMessage = new LocalizableResourceString(nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.PropertyNamesShouldNotMatchGetMethodsMessageBaseType), MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
         private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.PropertyNamesShouldNotMatchGetMethodsDescription), MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
-        
-        internal static DiagnosticDescriptor SameTypeRule = new DiagnosticDescriptor(RuleId,
+
+        internal static DiagnosticDescriptor DeclaringTypeRule = new DiagnosticDescriptor(RuleId,
```
Comment (mavasani) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PropertyNamesShouldNotMatchGetMethods.cs]: This seems like another case where FXCop reports diagnostics with same ID but multiple messages - see https://github.com/dotnet/roslyn-analyzers/pull/742#discussion_r46892144.

@morrisjoe you should replace multiple descriptors here with a single descriptor with a message format `"{0}"`, you can see how Nick did so https://github.com/dotnet/roslyn-analyzers/pull/742/files

Comment (mavasani) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PropertyNamesShouldNotMatchGetMethods.cs]: Filed https://github.com/dotnet/roslyn-analyzers/issues/745 to fix the stub generator.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PropertyNamesShouldNotMatchGetMethods.cs [hunk 3]
```diff
@@ -6,44 +6,112 @@
...
+
+            foreach (var type in types)
+            {
+                var membersFound = type.GetMembers(identifier);
+                if (membersFound != null && membersFound.Length > 0)
+                {
+                    foreach (var member in membersFound)
+                    {
+                        if ((symbol.Kind == SymbolKind.Property && member.Kind == SymbolKind.Method) ||
+                            (symbol.Kind == SymbolKind.Method && member.Kind == SymbolKind.Property && symbol.ContainingType != type))
```
Comment (srivatsn) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PropertyNamesShouldNotMatchGetMethods.cs]: why do you need the symbol.ContainingType != type check for just methods here?

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PropertyNamesShouldNotMatchGetMethods.cs (hunk 2 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PropertyNamesShouldNotMatchGetMethods.cs]: made changes to address this. re-submitted

File: src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PropertyNamesShouldNotMatchGetMethods.cs (hunk 3 above)
Comment (morrisjoe) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PropertyNamesShouldNotMatchGetMethods.cs]: Discussed off-line. Without that check, diagnostic will fire twice for the same issue

**Comment by srivatsn** [comment]
:+1:

**Pull request merged**